mozilla-release/browser/base/content/tabbrowser.xml

```diff
@@ -6511,9 +6525,7 @@
         <![CDATA[
           if (!this._linkedBrowser)
             return;
-          // Copied from browser/components/sessionstore/content/content-sessionStore.js
-          this._linkedBrowser.loadContext.QueryInterface(Ci.nsIDocShell).
-              addWeakPrivacyTransitionObserver(this);
+          this._linkedBrowser.loadContext.addWeakPrivacyTransitionObserver(this);
```

Comment:
As you've probably noticed, a <xul:browser> can flip remoteness. For example, if the user is viewing mozilla.org in a remote browser, and then goes to about:addons, the remoteness of the browser is flipped, the session history is brain-transplanted into the new non-remote browser, and the browser is sent to about:addons.

<xul:browser>'s will decide when attempting to load a URI whether or not it can be loaded in a remote or non-remote browser. That decision-making logic is here:

http://searchfox.org/mozilla-central/rev/f5c9e9a249637c9abd88754c8963ecb3838475cb/browser/modules/E10SUtils.jsm#28

And from the sounds of your email, you need to make your listeners resilient to this remoteness flipping.

Thankfully, we fire an event when we flip remoteness, which you can listen for, and then set up your observer again.

http://searchfox.org/mozilla-central/rev/f5c9e9a249637c9abd88754c8963ecb3838475cb/browser/base/content/tabbrowser.xml#1793